acvm-repo/acvm/src/compiler/optimizers/constant_backpropagation.rs

```diff
@@ -172,7 +175,13 @@ impl ConstantBackpropagationOptimizer {
                             Ok(BrilligSolverStatus::Finished) => {
                                 // Write execution outputs
                                 match solver.finalize(&mut known_witnesses, &new_brillig) {
-                                    Ok(()) => continue,
+                                    Ok(()) => {
+                                        // If we've managed to execute the brillig opcode at compile time, we can now just write in the
+                                        // results as constants for the rest of the circuit.
+                                        //
+                                        // TODO: what if these are required to be witnesses e.g. inputs to black box functions + array get/set?
```

Comment:
In `finalize` we write the brillig outputs to the witness? I am a bit confused by this TODO

Reply:
Oh I see nevermind

Reply:
The TODO is due to the fact that certain opcodes require that the input is a witness and so we want to ensure that we don't cause problems in this case. I need to have a check of this.

Reply:
Yeah makes sense. Either way it would probably be an easy fix where it can just be included in `required_witnesses`.